lib/tlscontext.c

```diff
@@ -335,6 +342,57 @@ _set_sni_in_client_mode(TLSSession *self)
   return FALSE;
 }
 
+void _write_line_to_keylog_file(const char *file_path, const char *line, FILE *keylog_file, GMutex *mutex)
+{
+  gint ret_val;
+  g_mutex_lock(mutex);
+  if(keylog_file)
+    {
+      ret_val = fprintf(keylog_file, "%s\n", line);
+      if (ret_val != strlen(line))
+        {
+          msg_error("Couldn't write to TLS keylogfile", evt_tag_errno("error", ret_val));
+        }
+    }
+  else
+    {
+      msg_error("Error opening keylog-file",
+                evt_tag_str(EVT_TAG_FILENAME, file_path),
+                evt_tag_error(EVT_TAG_OSERROR));
+    }
+  g_mutex_unlock(mutex);
+}
+
+static inline void
+_dump_tls_keylog(const SSL *ssl, const char *line)
+{
+  if(ssl != NULL)
+    {
+      const char *file_path = SSL_get_ex_data(ssl, _TLS_KEYLOG_INDEX);
+      GMutex *file_lock = SSL_get_ex_data(ssl, _TLS_KEYLOG_FILE_LOCK);
+      FILE *file = SSL_get_ex_data(ssl, _TLS_KEYLOG_FILE);
+      setbuf(file, NULL);
+      _write_line_to_keylog_file(file_path, line, file, file_lock);
+    }
+}
+
+static gboolean
+tls_session_keylog_setup_file(TLSSession *self, TLSContext *ctx)
+{
+  self->ctx->keylog_file = fopen(ctx->keylog_file_path, "a");
```

Comment:
After IRL discussion, we agreed that opening/closing the file inside the callback may be more expedient, as it is not a frequent operation in syslog-ng. We could also call `fflush()`, before closing the file, so a real-time packet capture with Wireshark is more likely to work.